resources/asciidoctor/spec/spec_helper.rb

```diff
@@ -47,25 +63,41 @@ def convert(input, extra_attributes = {}, warnings_matcher = eq(''))
 # Out:
 #   converted        - converted docbook text
 #   logs             - lines logged
-RSpec.shared_context 'convert' do
+RSpec.shared_context 'convert with logs' do
   let(:convert_logger) { Asciidoctor::MemoryLogger.new }
   let!(:converted) do
     # We use let! here to force the conversion because it populates the logger
-    attributes = {
-      'docdir' => File.dirname(__FILE__),
-    }
-    attributes.merge! convert_attributes if defined?(convert_attributes)
-    Asciidoctor.convert input,
-        safe:       :unsafe,  # Used to include "funny" files.
-        backend:    :docbook45,
-        logger:     convert_logger,
-        doctype:    :book,
-        attributes: attributes,
-        sourcemap:  true
+    extra_attributes = defined?(convert_attributes) ? convert_attributes : {}
+    internal_convert input, convert_logger, extra_attributes
   end
   let(:logs) do
     convert_logger.messages
       .map { |l| "#{l[:severity]}: #{l[:message].inspect}" }
       .join("\n")
   end
 end
+
+##
+# Converts asciidoc to docbook, asserting that nothing is logged during the
+# conversion.
+#
+# In:
+#   input            - asciidoc text to convert
+#   extra_attributes - attributes added to the conversion - defaults to {}
+#
+# Out:
+#   converted        - converted docbook text
+RSpec.shared_context 'convert no logs' do
+  let(:converted) do
+    convert_logger = Asciidoctor::MemoryLogger.new
+    extra_attributes = defined?(convert_attributes) ? convert_attributes : {}
+    converted = internal_convert input, convert_logger, extra_attributes
+    if convert_logger.messages.empty? == false
+      raise "Expected no logs but got:\n" +
+        convert_logger.messages
+          .map { |l| "#{l[:severity]}: #{l[:message].inspect}" }
+          .join("\n")
+    end
+    converted
+  end
+end
```

Comment:
This feels clean to me now! But I now feel like raising an error in this let block should be a `it` test instead.

Reply:
So I started there but that has problems when you include the context and then nest. So, like this:
```
context 'foo' do
  include_context 'convert no logs'
  context 'bar' do
    let(:input) { ... }
    it 'blah' { ... }
  end
  context 'baz' do
    let(:input) { ... }
    it 'blah' { ... }
  end
end
```

When you do the assertion with `it` then you only get the no-log assertion in the top level.

Reply:
ah, I see. The only way to solve that is to do the following:

```ruby
context 'foo' do
  context 'bar' do
   include_context 'convert no logs'
    let(:input) { ... }
    it 'blah' { ... }
  end
  context 'baz' do
    include_context 'convert no logs'
    let(:input) { ... }
    it 'blah' { ... }
  end
end
```

But maybe that's annoyingly redundant.

Reply:
My problem with that is that it is super easy to forget!

Reply:
carry on then :)

Reply:
> carry on then :)

Does that mean "I approve of this PR?" :smiling_imp:

Reply:
Indeed!